daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/TransactionPreprocessor.scala

```diff
@@ -95,28 +61,52 @@ private[preprocessing] final class TransactionPreprocessor(
       tx: GenTransaction[NodeId, Cid]
   ): (ImmArray[speedy.Command], Set[ContractId]) = {
 
-    type Acc = ((Set[Value.ContractId], Set[Value.ContractId]), BackStack[speedy.Command])
+    type Acc = (Set[Value.ContractId], Set[Value.ContractId], BackStack[speedy.Command])
 
-    val ((localCids, _), cmds) =
-      tx.roots.foldLeft[Acc](((Set.empty, Set.empty), BackStack.empty)) {
-        case ((cids, stack), id) =>
-          tx.nodes.get(id) match {
+    val (_, globalCids, cmds) =
+      tx.roots.foldLeft[Acc]((Set.empty, Set.empty, BackStack.empty)) {
+        case ((localCids0, globalCids0, stack), id) =>
+          val (newGlobals, newLocals, cmd) = tx.nodes.get(id) match {
             case None =>
               fail(s"invalid transaction, root refers to non-existing node $id")
             case Some(node) =>
               node match {
+                case create: Node.NodeCreate[Cid] =>
+                  val (cmd, newCids) =
+                    commandPreprocessor.unsafePreprocessCreate(create.templateId, create.arg)
+                  val newGlobalCids = newCids.filterNot(localCids0)
+                  (newGlobalCids, List(create.coid), cmd)
+                case exe: Node.NodeExercises[_, Cid] =>
+                  val templateId = exe.templateId
+                  val (cmd, newCids) =
+                    commandPreprocessor.unsafePreprocessExercise(
+                      templateId,
+                      exe.targetCoid,
+                      exe.choiceId,
+                      exe.chosenValue,
+                    )
+                  val newGlobalCids = newCids.filterNot(localCids0)
```

Comment:
Instead of filtering at each step, can we first accumulate everything and then drop the local cids from the global cids at the end? That seems easier to get right.

Reply:
nvm, the added check below wouldn’t work in that case.